src/charm.py

```diff
@@ -8,18 +8,157 @@
+        self._name = self.model.app.name
+
+        self.service_patcher = KubernetesServicePatch(
+            self, [("oathkeeper-api", OATHKEEPER_API_PORT)]
+        )
+
+        self.ingress = IngressPerAppRequirer(
+            self,
+            relation_name="ingress",
+            port=OATHKEEPER_API_PORT,
+            strip_prefix=True,
+        )
+
+        self.framework.observe(self.on.oathkeeper_pebble_ready, self._on_oathkeeper_pebble_ready)
+
+        self.framework.observe(self.ingress.on.ready, self._on_ingress_ready)
+        self.framework.observe(self.ingress.on.revoked, self._on_ingress_revoked)
+
+    @property
+    def _oathkeeper_layer(self) -> Layer:
+        """Returns a pre-configured Pebble layer."""
+        layer_config = {
+            "summary": "oathkeeper-operator layer",
+            "description": "pebble config layer for oathkeeper-operator",
+            "services": {
+                self._container_name: {
+                    "override": "replace",
+                    "summary": "Oathkeeper Operator layer",
+                    "command": f"oathkeeper serve -c {self._oathkeeper_config_path}",
+                    "startup": "enabled",
+                }
+            },
+            "checks": {
+                "alive": {
+                    "override": "replace",
+                    "http": {"url": f"http://localhost:{OATHKEEPER_API_PORT}/health/alive"},
+                },
+                "ready": {
+                    "override": "replace",
+                    "http": {"url": f"http://localhost:{OATHKEEPER_API_PORT}/health/ready"},
+                },
+            },
+        }
+        return Layer(layer_config)
+
+    @property
+    def _oathkeeper_service_is_created(self) -> bool:
+        try:
+            self._container.get_service(self._container_name)
+        except (ModelError, RuntimeError):
+            return False
+        return True
+
+    @property
+    def _kratos_login_url(self) -> str:
+        # This URL is a placeholder, it will be replaced by a juju integration
+        return f"http://kratos.{self.model.name}.svc.cluster.local:4433/self-service/login/browser"
+
+    def _render_access_rules_file(self) -> str:
+        """Render the access rules file."""
+        with open("templates/access-rules.yaml.j2", "r") as file:
+            template = Template(file.read())
+
+        rendered = template.render(
+            kratos_login_url=self._kratos_login_url,
+            return_to="http://default-url.com",
+        )
+        return rendered
+
+    def _render_conf_file(self) -> str:
+        """Render the Oathkeeper configuration file."""
+        with open("templates/oathkeeper.yaml.j2", "r") as file:
+            template = Template(file.read())
+
+        rendered = template.render(
+            kratos_session_url=f"http://kratos.{self.model.name}.svc.cluster.local:4433/sessions/whoami",
```

Comment:
Nitpick from a charming freshman:

- Is it better we group the two Kratos' exposed endpoints together somewhere in the `__init__` method?